### DISCUSSION TITLE
Sharing vertex attribute data between meshes

### DISCUSSION THREAD

**Discussion by aloucks** [bevyengine/bevy#2874] [Ideas]
Is it possible to share attribute data between multiple meshes? 

I'm thinking about this in the context of terrain where I may have many mesh chunks that all have the same indices and 2d vertex data. The terrain would be organized into blocks and each block would be subdivided into chunks. A block would have it's own heightmap and splatmap. A mesh chunk would have a unique set of UVs that are used to fetch the height/splat data from the corresponding block image in the vertex shader. Having the block subdivided into smaller chunks would allow for better frustum culling and enabling the splat lookup to use different palettes. 

From what I can tell, it doesn't seem like it is possible to share attribute data today. Ideally, I'd like to be able to de-duplicate this data both on the CPU side and on the GPU side.

Perhaps [VertexAttributeValues](https://docs.rs/bevy/0.5.0/bevy/render/mesh/enum.VertexAttributeValues.html)  could be altered to store `Arc<Vec<T>>` instead of  `Vec<T>` and [`Arc::ptr_eq`](https://doc.rust-lang.org/std/sync/struct.Arc.html#method.ptr_eq) could be used to determine if the GPU buffer should be shared?

**Comment by VVishion** [comment]
I can't really tell what your requirements are. But it looks like you could store your mesh chunks in `ResMut<Assets<Mesh>>` and reference/reuse them by their handle.

**Reply by aloucks** [reply]
Sorry, perhaps I wasn't clear. The issue is not related to being able to store the mesh data in general. The issue is that it's overly wasteful both in CPU and GPU memory. All of the vertex data and indices would be duplicated across all the meshes. The only attribute that's different between them would be the UVs.

**Reply by ebiggs** [reply]
Could you just put the uv offsets  in the instance data, then adjust them in the vertex shader?

**Comment by smessmer** [comment]
I have the same issue for pretty much the same use case. I am trying to build a tiled terrain system and each tile has the same triangle strip indices, but with different height values for the position vertices. To save GPU resources, I would like to share the index buffer between multiple Meshes.

There seems to be a `GpuMesh` type in bevy that allows manual setting of buffers, but I'm not sure if that can be used directly or if yes, how it could be used.

**Comment by Kees-van-Beilen** [comment]
Actually came to a situation where this was necessary. I'm going to write an implementation next weekend if possible. 

edit.

I did just notice you also mentioned that the UV should be unique per instance. This is not possible when sharing bufferdata.
